lib/srv/db/redis/engine.go

```diff
@@ -0,0 +1,460 @@
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.proxyConn = clientConn
+	e.clientReader = redis.NewReader(clientConn)
+	e.sessionCtx = sessionCtx
+
+	return nil
+}
+
+// authorizeConnection does authorization check for MongoDB connection about
+// to be established.
+func (e *Engine) authorizeConnection(ctx context.Context, sessionCtx *common.Session) error {
+	ap, err := e.Auth.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	mfaParams := services.AccessMFAParams{
+		Verified:       sessionCtx.Identity.MFAVerified != "",
+		AlwaysRequired: ap.GetRequireSessionMFA(),
+	}
+
+	dbRoleMatchers := role.DatabaseRoleMatchers(
+		sessionCtx.Database.GetProtocol(),
+		sessionCtx.DatabaseUser,
+		// Leave database empty as Redis integration doesn't support schema access control.
+		"",
+	)
+	err = sessionCtx.Checker.CheckAccess(
+		sessionCtx.Database,
+		mfaParams,
+		dbRoleMatchers...,
+	)
+	if err != nil {
+		e.Audit.OnSessionStart(e.Context, sessionCtx, err)
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// SendError sends error message to connected client.
+func (e *Engine) SendError(redisErr error) {
+	if redisErr == nil || utils.IsOKNetworkError(redisErr) {
+		return
+	}
+
+	if err := e.sendToClient(redisErr); err != nil {
+		e.Log.Errorf("failed to send message to the client: %v", err)
+		return
+	}
+}
+
+// sendToClient sends a command to connected Redis client.
+func (e *Engine) sendToClient(vals interface{}) error {
+	buf := &bytes.Buffer{}
+	wr := redis.NewWriter(buf)
+
+	if err := writeCmd(wr, vals); err != nil {
+		return trace.BadParameter("failed to convert error to a message: %v", err)
+	}
+
+	if _, err := e.proxyConn.Write(buf.Bytes()); err != nil {
+		return trace.ConnectionProblem(err, "failed to send message to the client")
+	}
+
+	return nil
+}
+
+// HandleConnection is responsible for connecting to a Redis instance/cluster and
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	// Check that the user has access to the database.
+	err := e.authorizeConnection(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err, "error authorized database access")
+	}
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	connectionOptions, err := ParseRedisURI(sessionCtx.Database.GetURI())
+	if err != nil {
+		return trace.BadParameter("Redis connection string is incorrect: %v", err)
+	}
+
+	var (
+		redisConn      redis.UniversalClient
+		connectionAddr = fmt.Sprintf("%s:%s", connectionOptions.address, connectionOptions.port)
+	)
+
+	// TODO(jakub): Use system CA bundle if connecting to AWS.
+	// TODO(jakub): Investigate Redis Sentinel.
+	if connectionOptions.cluster {
+		redisConn = redis.NewClusterClient(&redis.ClusterOptions{
+			Addrs:     []string{connectionAddr},
+			TLSConfig: tlsConfig,
+		})
+	} else {
+		redisConn = redis.NewClient(&redis.Options{
```

Comment:
any other options that we are interested to set? like `PoolSize`, timeouts, etc.

Reply:
I mentioned that above in my comment. I think that some options can be useful, here is a full list https://github.com/go-redis/redis/blob/master/options.go#L60-L105

I don't mind adding some of those now or after v9.

@r0mant @smallinsky Thoughts?

Reply:
I would pick some reasonable defaults keeping in mind that we're most likely proxying a connection from an interactive client and not an application.

Does Redis send any client options as a part of the protocol? For SQL Server for example I pass all client options to the server and then all server options back to the client but I don't know if we can do something similar here.

Reply:
Redis itself doesn't have any options that we can set. All timeouts, pool size, retries are implemented on the client side (go-redis). In my opinion defaults in go-redis are fine for CLI usage (at least I've never problem with them).